android/src/org/kiwix/kiwixmobile/utils/files/FileSearch.java

```diff
@@ -48,9 +49,66 @@
   // Array of zim file extensions
   public static final String[] zimFiles = {"zim", "zimaa"};
 
+
+  private final Context context;
+  private final ResultListener listener;
+
+  private boolean fileSystemScanCompleted = false;
+  private boolean mediaStoreScanCompleted = false;
+
+  public FileSearch(Context ctx, ResultListener listener) {
+    this.context = ctx;
+    this.listener = listener;
+  }
+
+  public void scan() {
+    fileSystemScanCompleted = false;
+    mediaStoreScanCompleted = false;
+
+    new Thread(new Runnable() {
+      @Override
+      public void run() {
+        scanFileSystem();
+        fileSystemScanCompleted = true;
+        checkCompleted();
+      }
+    }).start();
+
+    new Thread(new Runnable() {
+      @Override
+      public void run() {
+        scanMediaStore();
+        mediaStoreScanCompleted = true;
+        checkCompleted();
+      }
+    }).start();
+  }
+
+  private synchronized void checkCompleted() {
+    if (mediaStoreScanCompleted && fileSystemScanCompleted)
+      listener.onScanCompleted();
+  }
+
+  public void scanMediaStore() {
+    ContentResolver contentResolver = context.getContentResolver();
+    Uri uri = MediaStore.Files.getContentUri("external");
+
+    String[] projection = {MediaStore.MediaColumns.DATA};
+    String selection = MediaStore.MediaColumns.DATA + " like ? or "+MediaStore.MediaColumns.DATA +" like ? ";
+
+    Cursor query = contentResolver.query(uri, projection, selection, new String[]{"%."+zimFiles[0], "%."+zimFiles[1]}, null);
+
+    try {
+      while (query.moveToNext()) {
+        onFileFound(query.getString(0));
```

Comment:
Mediascanner can sometimes return phantom files as it doesn't update instantly. We should check the file really exists before adding it.